File: frontend/src/services/seriesService.js
/*API*/
import api from "./api";

/*SERIES*/
export async function getPopularSeries() {
  const response = await api.get("/series/popular");
  return response.data.results;
}

export async function getTrendingSeries() {
  const response = await api.get("/series/trending");
  return response.data.results;
}

export async function searchSeries(query, page = 1) {
    const { data } = await api.get("/series/search", {
        params: {
            query,
            page
        }
    });
    return data;
}

export async function getNewSeriesReleases() {
  const response = await api.get("/series/newReleases");
  return response.data.results;
}

export async function getRandomSeries() {
  const response = await api.get("/series/random");
  return response.data;
}

export async function getSeriesGenres() {
  const response = await api.get("/genres/series");  
  return response.data;  
}

export async function discoverSeries(filters, page) {

    const { data } = await api.get("/series/discover", {
        params: {
            page,
            genre: filters.genre || undefined,
            yearFrom: filters.yearFrom || undefined,
            yearTo: filters.yearTo || undefined,
            rating: filters.rating !== "all" ? filters.rating : undefined,
            sortBy: filters.sortBy || undefined,
        },
    });

    return data;
}

export async function getTrailer(mediaId, mediaType) {
  const { data } = await api.get(`/${mediaType}/${mediaId}/trailer`);
  return data;
}

export async function getSeriesDetails(movieId) {
    const { data } = await api.get(`/series/${movieId}/details`);
    return data;
}

File: backend/src/main/java/com/example/backend/service/SeriesService.java
package com.example.backend.service;

import com.example.backend.tmdb_client.client.SeriesClient;
import com.example.backend.tmdb_client.dto.series.BasicSeriesDTO;
import com.example.backend.tmdb_client.dto.series.SeriesDetailsDTO;
import com.example.backend.tmdb_client.dto.series.SeriesTrailerDTO;
import com.example.backend.tmdb_client.response.series.*;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class SeriesService {

    private final SeriesClient seriesClient;

    public SeriesListResponse getPopularSeries() {
        return seriesClient.getPopularSeries();
    }

    public SeriesListResponse getTrendingSeries() { return seriesClient.getTrendingSeries(); }

    public NewReleaseSeriesListResponse getNewReleaseSeries() { return seriesClient.getNewSeriesReleases(); }

    public BasicSeriesDTO getRandomMovie(){
        return seriesClient.getRandomSeries().getResults().get(0);
    }

    public SearchSeriesListResponse getSearchSeriesResults(String query, int page) { return seriesClient.getSearchSeriesResults(query,page); }

    public DiscoverSeriesListResponse discoverSeries(Integer page, Integer genre, Integer yearFrom, Integer yearTo, Double rating, String sortBy) {
        return seriesClient.discoverSeries(page, genre, yearFrom, yearTo, rating, sortBy);
    }

    public SeriesTrailerDTO getSeriesTrailer(Integer seriesId) {

        SeriesTrailerResponse response = seriesClient.getSeriesTrailer(seriesId);

        if (response == null || response.getResults() == null) {
            return null;
        }

        return response.getResults().stream()
                .filter(video -> "YouTube".equals(video.getSite()))
                .filter(video -> "Trailer".equals(video.getType()))
                .filter(video -> Boolean.TRUE.equals(video.getOfficial()))
                .findFirst()

                .or(() ->
                        response.getResults().stream()
                                .filter(video -> "YouTube".equals(video.getSite()))
                                .filter(video -> "Trailer".equals(video.getType()))
                                .findFirst()
                )

                .or(() ->
                        response.getResults().stream()
                                .filter(video -> "YouTube".equals(video.getSite()))
                                .findFirst()
                )

                .orElse(null);
    }

    public SeriesDetailsDTO getSeriesDetails(Integer seriesId) {

        SeriesDetailsResponse response = seriesClient.getSeriesDetails(seriesId);

        return SeriesDetailsDTO.builder()

                .id(response.getId())
                .originalName(response.getOriginalName())
                .overview(response.getOverview())
                .runtime(response.getRuntime())
                .firstAirDate(response.getFirstReleaseDate())
                .rating(response.getVoteAverage())
                .posterPath(response.getPosterPath())
                .totalEpisodes(response.getTotalEpisodes())
                .totalSeasons(response.getTotalSeasons())
                .status(response.getStatus())
                .genres(response.getGenres())
                .productionCompanies(response.getProductionCompanies())
                .productionCountries(response.getProductionCountries())
                .spokenLanguages(response.getSpokenLanguages())
                .build();
    }
}
